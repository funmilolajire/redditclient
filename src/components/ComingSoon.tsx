import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';

export const ComingSoon = () => {
    const boxRef = useRef<HTMLDivElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({ repeatDelay: 2 })
        tl.from([boxRef.current], {
            y: '-55vh',
            duration: 3,
            delay: 2,
            opacity: 0,
            ease: 'elastic'
        })
    })

    return (
        <div ref={boxRef} className="container flex items-center justify-center h-screen font-extrabold ComingSoon">
            <div className="bg-yellow-500 border-2 border-gray-800 border-solid skew-y-2 transform-gpu text-gray-50 text-9xl">
                <p>Coming Soon</p>
            </div>
        </div>
    )
}