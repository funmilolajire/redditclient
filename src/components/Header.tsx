import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';
import logo from '../images/logo192.png';

export const Header = () => {
    const boxRef = useRef<HTMLDivElement>(null);
    const logoRef = useRef<HTMLImageElement>(null);
    const titleRef = useRef<HTMLHeadingElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({});
        tl.from([boxRef.current], {
            duration: 1
        })
        tl.from([logoRef.current], {
            x: '-3000px',
            opacity: 0,
            ease: 'circ'
        })
        tl.from([titleRef.current], {
            x: '3000px',
            opacity: 0,
            ease: 'slow'
        })
    })

    return (
        <header>
            <div className="header-inner">
                <section ref={boxRef}>
                    <figure ref={logoRef}>
                        <img src={logo} alt="site logo" />
                    </figure>
                    <h1 ref={titleRef}>
                        <span>Reddit</span>Client
                    </h1>
                </section>
            </div>
        </header>
    )
}