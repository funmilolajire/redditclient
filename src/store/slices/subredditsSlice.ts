import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { subredditsList } from '../api'

//set initial state
const initialState: SubredditState = {
    value: [],
    loading: true
}

//get subreddits list
export const getSubredditsList = createAsyncThunk('subreddits', subredditsList)

//create subreddits slice
const subredditsSlice = createSlice({
    name: "subreddits",
    initialState,
    reducers: {},
    extraReducers: builder => {
        builder.addCase(getSubredditsList.fulfilled, (state: SubredditState, { payload }) => {
            if (payload)
                state.value = payload;
            state.loading = false;
        })
    }
})

//export subreddits slice reducer
export default subredditsSlice.reducer