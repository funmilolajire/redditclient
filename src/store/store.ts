import { configureStore } from '@reduxjs/toolkit';
import subredditsReducer from './slices/subredditsSlice';
import postsReducer from './slices/postsSlice';
import moreInformationReducer from './slices/moreInformationSlice';

export const store = configureStore({
  reducer: {
    subreddits: subredditsReducer,
    posts: postsReducer,
    moreInformation: moreInformationReducer
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch