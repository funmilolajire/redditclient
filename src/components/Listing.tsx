import { useRef, useLayoutEffect, useCallback } from 'react';
import { useAppDispatch, useAppSelector } from '../store/hooks';
import { handleListingChange, handleTermChange } from '../store/slices/postsSlice';

export const Listing = ({ listItem }: { listItem: string }) => {
    const listingRef = useRef<HTMLLIElement>(null);

    //handle listing change and active class
    const dispatch = useAppDispatch();
    const listing = useAppSelector(state => state.posts.listing);
    const handleListing = () => {
        if (null !== listingRef.current && listing !== listingRef.current.id) {
            let newListing = listingRef.current.id;
            dispatch(handleListingChange(newListing))
            dispatch(handleTermChange(''))
        }
    }
    const handleActiveClass = useCallback(() => {
        if (null !== listingRef.current) {
            const classList = listing === listingRef.current.id ? "Listing active" : "Listing";
            listingRef.current.className = classList
        }
    }, [listing])
    useLayoutEffect(() => {
        handleActiveClass();
    }, [handleActiveClass])


    //JSX
    return (
        <li ref={listingRef} data-testid={listItem} onClick={handleListing} id={listItem}>
            {listItem}
        </li>
    )
}