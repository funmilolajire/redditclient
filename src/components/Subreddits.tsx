import { useRef, useLayoutEffect, useEffect, useCallback } from 'react';
import { useAppDispatch, useAppSelector } from '../store/hooks';
import { getSubredditsList } from '../store/slices/subredditsSlice';
import { handleSubredditChange } from '../store/slices/postsSlice';
import { gsap } from 'gsap';
import { Subreddit } from './Subreddit';
import { IoIosClose } from 'react-icons/io';

export const Subreddits = () => {
    //animations
    const subRef = useRef<HTMLHeadingElement>(null);
    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([subRef.current], {
            delay: -1,
            scale: 2,
            x: '-100%',
            duration: 2,
            ease: 'elastic',
        })
    }, [])

    //get subreddits list
    const subreddits = useAppSelector((state: { subreddits: { value: string[] } }): string[] => state.subreddits.value)
    const dispatch = useAppDispatch();
    const getSubreddits = useCallback(async () => {
        await dispatch(getSubredditsList());
    }, [dispatch])
    useEffect(() => {
        getSubreddits();
    }, [getSubreddits])

    //get and remove active subreddit
    const subreddit = useAppSelector(state => state.posts.subreddit)
    const removeSubreddit = () => {
        dispatch(handleSubredditChange(''))
    }

    //JSX
    return (
        <section className="pt-12 Subreddits">
            <h2 ref={subRef} className="">subreddits</h2>
            {subreddit.length !== 0 && <div className="current--subreddit"><span className="subreddit">{subreddit}</span><span onClick={removeSubreddit} className="icon"><IoIosClose /></span></div>}
            <ul className="subreddits--list">
                {subreddits && subreddits.map(subreddit => <Subreddit key={subreddit}>{subreddit}</Subreddit>)}
            </ul>
        </section>
    )
}