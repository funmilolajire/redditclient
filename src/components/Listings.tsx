import { useRef, useLayoutEffect } from 'react';
import { Listing } from './Listing';
import { gsap } from 'gsap';

export const Listings = () => {
    const listings: string[] = ['hot', 'new', 'top', 'best', 'rising', 'controversial'];

    //animations
    const listingBoxRef = useRef<HTMLDivElement>(null);
    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([listingBoxRef.current], {
            x: '1200%',
            delay: 2,
            duration: 2,
            ease: 'power3.inOut'
        })
    }, [])

    return (
        <div ref={listingBoxRef} className="Listings">
            <ul className="Listings--List">
                {
                    listings && listings.map(listing => <Listing key={listing} listItem={listing} />)
                }
            </ul>
        </div>
    )
}