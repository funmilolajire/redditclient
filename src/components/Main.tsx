import { Listings } from './Listings';
import { Posts } from './Posts';
export const Main = () => (
    <div className="Main">
        <Listings />
        <Posts />
    </div>
)