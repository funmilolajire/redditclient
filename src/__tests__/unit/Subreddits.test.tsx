import { mount } from 'enzyme';
import { Subreddits } from '../../components/Subreddits';
import { Provider } from 'react-redux';
import { store } from '../../store/store';
import { getSubredditsList } from '../../store/slices/subredditsSlice';

describe('renders Subreddits component', () => {
    store.dispatch(getSubredditsList)
    const subreddits = store.getState().subreddits.value;

    const SubredditsComponent = mount(
        <Provider store={store}>
            <Subreddits />
        </Provider>
    )
    test('make a snapshot of Subreddits Component', () => {
        expect(SubredditsComponent).toMatchSnapshot()
    });

})