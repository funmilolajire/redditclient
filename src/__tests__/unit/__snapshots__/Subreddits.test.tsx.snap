// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Subreddits component make a snapshot of Subreddits Component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Subreddits>
    <section
      className="pt-12 Subreddits"
    >
      <h2
        className=""
      >
        subreddits
      </h2>
      <ul
        className="subreddits--list"
      />
    </section>
  </Subreddits>
</Provider>
`;
